src/app/views/canvas/mark_editor.tsx

```diff
@@ -283,39 +267,73 @@ export class SingleMarkView
               bboxCircle.cy + bboxCircle.radius
             ];
           }
+          case "line": {
+            const bboxLine = bbox as Prototypes.BoundingBox.Line;
+            xBounds = [bboxLine.x1, bboxLine.x2];
+            yBounds = [bboxLine.y1, bboxLine.y2];
+          }
+        }
+        if (xBounds.length > 0) {
+          // y is the same size
+          boundingRects.push([
+            Math.min(...xBounds),
+            Math.max(...xBounds),
+            Math.min(...yBounds),
+            Math.max(...yBounds)
+          ]);
         }
-        x1 = Math.min(x1, ...xBounds);
-        x2 = Math.max(x2, ...xBounds);
-        y1 = Math.min(y1, ...yBounds);
-        y2 = Math.max(y2, ...yBounds);
       }
     }
 
-    const cx = (x1 + x2) / 2;
-    const cy = (y1 + y2) / 2;
-    const overshoot = 0.4;
-    const scale1 = width / (1 + Math.abs(x2 - x1) * (1 + overshoot));
-    const scale2 = height / (1 + Math.abs(y2 - y1) * (1 + overshoot));
-    const scale = Math.min(scale1, scale2);
-    const zoom = {
-      centerX: width / 2 - cx * scale,
-      centerY: height / 2 + cy * scale,
-      scale
-    } as ZoomInfo;
-    return zoom;
+    // If there's no bounding rect found
+    if (boundingRects.length == 0) {
+      const cx = 0;
+      const cy = 0;
+      const { x1, x2, y1, y2 } = glyphState.attributes as {
+        x1: number;
+        y1: number;
+        x2: number;
+        y2: number;
+      };
+      const overshoot = 0.4;
+      const scale1 = width / (1 + Math.abs(x2 - x1) * (1 + overshoot));
+      const scale2 = height / (1 + Math.abs(y2 - y1) * (1 + overshoot));
+      const scale = Math.min(scale1, scale2);
+      const zoom = {
+        centerX: width / 2 - cx * scale,
+        centerY: height / 2 + cy * scale,
+        scale
+      } as ZoomInfo;
+      return zoom;
+    } else {
+      const x1 = Math.min(...boundingRects.map(b => b[0]));
+      const x2 = Math.max(...boundingRects.map(b => b[1]));
+      const y1 = Math.min(...boundingRects.map(b => b[2]));
+      const y2 = Math.max(...boundingRects.map(b => b[3]));
+      const cx = (x1 + x2) / 2;
+      const cy = (y1 + y2) / 2;
+      const overshoot = 0.4;
+      const scale1 = width / (1 + Math.abs(x2 - x1) * (1 + overshoot));
+      const scale2 = height / (1 + Math.abs(y2 - y1) * (1 + overshoot));
+      const scale = Math.min(scale1, scale2);
+      const zoom = {
+        centerX: width / 2 - cx * scale,
+        centerY: height / 2 + cy * scale,
+        scale
+      } as ZoomInfo;
+      return zoom;
+    }
   }
 
   public doAutoFit() {
     const newZoom = this.getFitViewZoom(this.props.width, this.props.height);
     if (!newZoom) {
       return;
     }
-    const isFocusing =
-      this.props.store.parent.currentSelection instanceof MarkSelection;
-    if (this.state.autoFitNextUpdate || !isFocusing) {
+    if (this.autoFitNextNUpdates > 0) {
```

Comment:
Seems like it could be brittle, why not just compare `this.state.zoom` to `newZoom`?

Reply:
yeah, that's a ugly solution, but the thing is that this function will be called twice after the mark is added (that's why `autoFitNextNUpdates` is set to 2). Once adding a mark, a selection event is trigged (update 1), and then the solver runs asynchronously, which later triggers update 2. If we auto fit to update 1, even though this.state.zoom will be different from newZoom, it won't look correct because the solver hasn't updated the states yet. Other thing is we don't want to do the auto scaling on everything, that would be more annoying than useful.

I think a better solution would be to make sure it doesn't run the auto fit on the selection event. I'll give it a shot.